subsys/bluetooth/host/l2cap.c

```diff
@@ -1942,47 +2019,37 @@ static int l2cap_chan_le_send(struct bt_l2cap_le_chan *ch,
 	/* Save state so it can be restored if we failed to send */
 	net_buf_simple_save(&buf->b, &state);
 
-	if ((buf->len <= ch->tx.mps) &&
-	    (net_buf_headroom(buf) >= BT_L2CAP_BUF_SIZE(0))) {
-		LOG_DBG("len <= MPS, not allocating seg for %p", buf);
-		seg = net_buf_ref(buf);
+	seg = get_seg(buf, ch->tx.mps, ch);
 
-		len = seg->len;
-	} else {
-		LOG_DBG("allocating segment for %p (%u bytes left)", buf, buf->len);
-		seg = l2cap_alloc_seg(ch);
-		if (!seg) {
-			LOG_DBG("failed to allocate seg for %p", buf);
-			atomic_inc(&ch->tx.credits);
-
-			return -EAGAIN;
-		}
-
-		/* Don't send more than TX MPS */
-		len = MIN(net_buf_tailroom(seg), ch->tx.mps);
-
-		/* Limit if original buffer is smaller than the segment */
-		len = MIN(buf->len, len);
-
-		net_buf_add_mem(seg, buf->data, len);
-		net_buf_pull(buf, len);
+	CHECKIF(!seg) {
+		/* Future work: Give the channel a tx state
+		 * machine, so that we remember that we took a
+		 * credit and don't need to give it back here.
+		 */
+		LOG_WRN("Out of segment buffers.");
+		atomic_inc(&ch->tx.credits);
+		return -ENOBUFS;
```

Comment:
Because the number of `seg_pool` is `CONFIG_BT_MAX_CONN`, and the number of dynamically created L2cap channels is uncertain (depends on the application), I think there is a certain probability of not getting seg buffer.

Is it possible to change -ENONUFS to -EAGAIN? Then it can be tried later. The premise for such changes is that the l2cap can be resumed in callback `l2cap_chan_seg_sent`.

Or, is it possible to integrate the processing of function `l2cap_chan_tx_process` into `bt_conn_tx_processor`?

Reply:
You're right, I had on my todo-list to test with a pool of size 1, but forgot to do it.
I'll make it work.

Reply:
Addressed by https://github.com/zephyrproject-rtos/zephyr/pull/72090/commits/37cfd30a28e83db59a94e09fd62c1de29ba5eea0